app/models/mixins/custom_actions_mixin.rb

```diff
@@ -0,0 +1,37 @@
+module CustomActionsMixin
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :custom_button_sets, :as => :owner, :dependent => :destroy
+    virtual_has_many :custom_buttons
+    virtual_has_one :custom_actions, :class_name => "Hash"
+    virtual_has_one :custom_action_buttons, :class_name => "Array"
+  end
+
+  def custom_buttons
+    generic_custom_buttons.select { |button| button.parent.nil? } + direct_custom_buttons
+  end
+
+  def custom_actions
+    generic_button_group = generic_custom_buttons.select { |button| !button.parent.nil? }
+    custom_button_sets_with_generics = custom_button_sets + generic_button_group.map(&:parent).uniq.flatten
+    {
+      :buttons       => custom_buttons.collect(&:expanded_serializable_hash),
+      :button_groups => custom_button_sets_with_generics.collect do |button_set|
+        button_set.serializable_hash.merge(:buttons => button_set.children.collect(&:expanded_serializable_hash))
+      end
+    }
+  end
+
+  def custom_action_buttons
+    custom_buttons + custom_button_sets.collect(&:children).flatten
+  end
+
+  def direct_custom_buttons
+    CustomButton.buttons_for(self).select { |b| b.parent.nil? }
+  end
+
+  def generic_custom_buttons
```

Comment:
@abellotti just updated with this to make the contract between this mixin and the host more explicit. It might be nice to add some lower-level tests for Vms now that this has been included - LMK if you'd rather I do that here or in a follow up. Thanks!